QED: now particles initialize a real component (optical depth) to be used for QED modules

lucafedeli88 opened the pull request:
This is a step towards the integration of the QED modules
The code should be compiled with QED=TRUE for these changes to take effect.

-  if "do_qed=1" is set for a given species in the inputfile  a real field named "tau" is added by in PhysicalParticleContainer

- if "do_qed_quantum_sync=1" AND "do_qed=1" are set for a given species in the inputfile  
the tau component is initialised by the Quantum Synchrotron engine. This is done in PhysicalParticleContainer. In principle there should be a check on the particle type, since only electrons and positrons should undergo this physical process. We are not doing any test for the moment

- if "do_qed_breit_wheeler =1" AND "do_qed=1" are set for a given species in the inputfile  
the tau component is initialised by the Breit Wheeler engine. This is done in PhotonParticleContainer,  
since this process concerns only photons.

There are also some wrappers for the QED modules developed in PICSAR

lucafedeli88 commented on Examples/Physics_applications/QED/inputs.2d:
```diff
@@ -57,6 +61,10 @@ electrons.uz_m = 500.
 electrons.uy_th = 2.
 electrons.uz_th = 50.
+##########QED####################
+electrons.do_qed = 1
```
The idea here is to have a flag to switch qed effects on/off for a given species.
This behaviour could be changed anyway (e.g. it could make sense to have a flag to switch off all QED effects for all the species)

lucafedeli88 commented on Examples/Physics_applications/QED/inputs.2d:
```diff
@@ -57,6 +61,10 @@ electrons.uz_m = 500.
 electrons.uz_th = 50.
+##########QED####################
+electrons.do_qed = 1
+electrons.do_qed_quantum_sync = 1
```
IF do_qed=1 this will enable Quantum Synchrotron process for the species. 
In principle we should check if the species is either electrons or positrons, but we are not doing that now.
One question: have you already implemented classical radiation reaction somewhere or is someone working on it?

lucafedeli88 replied:
Ok! I could do that then, since it is even more urgent than QED (I've opened a " dedicated issue")

lucafedeli88 commented on Examples/Physics_applications/QED/inputs.2d:
```diff
@@ -78,6 +86,10 @@ photons.uz_m = 500.
 photons.ux_th = 2.
 photons.uy_th = 2.
 photons.uz_th = 50.
+##########QED####################
```
Here do_qed_breit_wheeler enables the Breit Wheeler process, which concerns only photons. This process is implemented only for PhotonParticleContainers, so there is no risk of doing something unphysical. However it could be beneficial for the user to receive an error message if he/she tries to enable this process for electrons...

lucafedeli88 replied:
Where would you suggest to add that? PhotonParticlesContainers are also PhysicalParticleContainers and they call the constructor of the base class... I could easily add a test in PhotonParticlesContainer to check for the do_qed_quantum_sync flag however

lucafedeli88 replied:
Where would you add that? PhotonParticleContainers are also PhysicalParticleContainers and they call the constructor of the base class. The easy thing (just pushed) is to add a check for  do_qed_quantum_sync in the constructor of PhotonParticleContainer

MaxThevenet replied:
Yeah, I see the issue. If you want to throw an error when a `PhysicalParticleContainer` that is not a `PhotonParticleContainer` has `do_qed_breit_wheeler=1`, we could add a new attribute in the `PhysicalParticleContainer`, like `is_photon`, and set it to the appropriate value. Then we could add this piece of code into the `PhysicalParticleContainer` constructor
```c++
bool do_qed_breit_wheeler;
pp.query("do_qed_breit_wheeler", do_qed_breit_wheeler);
if (is_photon == 0){
    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
        do_qed_breit_wheeler == 0,
        "ERROR: do_qed_breit_wheeler can be 1 for species listed in particles.photon_species only!");
}
```

lucafedeli88 commented on Source/Diagnostics/ParticleIO.cpp:
```diff
@@ -112,6 +112,12 @@ MultiParticleContainer::WritePlotFile (const std::string& dir) const
                 int_flags.resize(1, 1);
             }
 
+#ifdef WARPX_QED
```
If QED effects are enabled for a species, we add a real "optical depth component" .

lucafedeli88 commented on Source/Particles/MultiParticleContainer.H:
```diff
@@ -22,6 +22,11 @@
 // concrete class derived from WarpXParticlContainer.
 //
 
+#ifdef WARPX_QED
```
breit_wheeler_engine_wrapper and quantum_sync_engine_wrapper are wrappers for the engines developed in PICSAR. During the simulation run we will use only calls to static functions of these classes. However, somewhere in the code, we'll need to initialize or load some lookup tables. Which is the most appropriate part of the code to do that?

lucafedeli88 replied:
Once in a simulation. So MultiParticleContainer::InitData is definitively the most appropriate place in the code. I'll do that (probably) in the next PR

lucafedeli88 commented on Source/Particles/PhotonParticleContainer.cpp:
```diff
@@ -21,12 +21,27 @@ using namespace amrex;
+
+    ParmParse pp(species_name);
+
+#ifdef WARPX_QED
```
An alternative here could be:
if do_qed_breit_wheeler=1, then QED processes are enabled for a given species (i.e. do_qed=1)

lucafedeli88 replied:
I would remove the compilation flag. I don't think that compilation with QED effects is significantly longer than the compilation without QED effects. However, we could do tests and we could discuss that in detail. However, for the moment, I would keep the compilation flag, since QED effects require the picsar library to be on a specific branch (QED)

lucafedeli88 replied:
Update: I've done a quick test. 
time make DIM=2                       --> 5m58,979s
time make DIM=2 QED=TRUE --> 6m29,229s
So, there seems to be  a significant difference (~ 8 %), though not huge.

MaxThevenet replied:
Ok, let's keep all tests.
And thanks for the timers!

lucafedeli88 commented on Source/Particles/PhysicalParticleContainer.cpp:
```diff
@@ -46,25 +46,45 @@ PhysicalParticleContainer::PhysicalParticleContainer (AmrCore* amr_core, int isp
+    //TODO: SHOULD CHECK IF SPECIES IS EITHER ELECTRONS OR POSITRONS!!
+#endif
+
+    //variable to set plot_flags size
```
I've changed a little bit how plot_flag_size is calculated. I think that the code is a bit cleaner then before...

lucafedeli88 replied:
ok

lucafedeli88 replied:
done. It's this PR: lucafedeli88-QED-update-doc

MaxThevenet replied:
As discussed in PR https://github.com/ECP-WarpX/WarpX/pull/357, let's just always dump this variable if QED is on.

lucafedeli88 commented on Source/QED/amrex_rng_wrapper.cpp:
```diff
@@ -0,0 +1,20 @@
+//This file provides a wrapper aroud the RNG
```
The QED library developed in PICSAR requires a "random number generator"
the only requirement is that it should provide
- a unf (a,b) function to generate uniformly distributed numbers between a and b
- an exp(l) function, to generate exponentially distributed numbers

lucafedeli88 replied:
Actually, it seems to me that `amrex::RandomNormal` generates a random variable distributed according to a Gaussian distribution. I need exp(l) to generate numbers distributed according to an exponential distribution. It could be possible to do a pull request to implement  `amrex::RandomExp` in AMReX. However, care should be taken since curand does not implement the exponential distribution, as far as I know.

> Could you give it a more human-readable name, e.g., `RandomUniform` ?

Yes, but it will take time, since it is the name used in the QED module of the PICSAR library

>Does it have anything specific to the BW engine, or could it be used somewhere else? In the latter case, could you remove mentions to BW in this file, and move the file to Utils?

I'll try to make this point clearer. Sorry for not having done that before.
All the engines (Breit Wheeler, Quantum Synchrotron and Schwinger) in the QED module of PICSAR are template classes:
```
     template<typename _REAL, class _RNDWRAP>
     class breit_wheeler_engine

     template<typename _REAL, class _RNDWRAP>
     class quantum_synchrotron_engine

     template<typename _REAL, class _RNDWRAP>
     class schwinger_pair_engine
```

_RNDWRAP should be a class able to provide the 'unf' and 'exp' functions. 
Actually, this is only needed if the engines are instantiated and used as objects
to update the optical depth and to calculate the outcome of a QED process. 
At present, generating the lookup tables requires to instantiate the engine, but
the RNG is not actually invoked. Although the engine classes could be instantiated on 
a GPU, we agreed to use only calls to static functions in GPU kernels. These calls do
not invoke the RNG, since a random number is passed as an argument. 
In principle I could get rid of this class entirely, rely only on static memeber function
calls in GPU kernels and do something like the following to generate the lookup tables:

```
class rng_dummy{};

breit_wheeler_engine<amrex::Real, rng_dummy> bw_engine;

```

Do you think that this solution could be preferable? 
In any case, my impression is that this solution would benefit from an intermediate layer (e.g. a wrapper class) between the QED library and WarpX. Please see my comment specifically on this point.

MaxThevenet replied:
> Could you give it a more human-readable name, e.g., RandomUniform ?
> > Yes, but it will take time, since it is the name used in the QED module of the PICSAR library
Ok sorry I was confused about that. No need to change the names (in particular if we get rid of this class, as you suggest).

Ok thanks for clarifying that point. Yes, I think that if you won't need the random number generator, we can get rid of this class and pass a dummy RNG class to the initialization.

And in any case, yes, let's keep a wrapper layer to make future changes easier :)

lucafedeli88 commented on Source/QED/breit_wheeler_engine_wrapper.cpp:
```diff
@@ -0,0 +1,16 @@
+AMREX_GPU_HOST_DEVICE
+warpx_breit_wheeler_engine::get_optical_depth(){
+    return internal_get_optical_depth(amrex::Random());
+}
```
this (static) function is a wrapper around the static function internal_get_optical_depth of the base class.
We might want to change the name, because get_optical_depth() is also the name of a non static member function of the base class. I could replace it with "optical_depth".
Actually internal_get_optical_depth does a very simple thing:

```
template<typename _REAL, class _RNDWRAP>
PXRMP_GPU
PXRMP_FORCE_INLINE
_REAL picsar::multi_physics::breit_wheeler_engine<_REAL, _RNDWRAP>::
internal_get_optical_depth(_REAL unf_zero_one_minus_epsi)
{
    return -log(one - unf_zero_one_minus_epsi);
}
```

lucafedeli88 replied:
thanks, I'll keep it as it is for the moment

lucafedeli88 commented on Source/QED/qed_wrapper_commons.h:
```diff
@@ -0,0 +1,13 @@
+
+#include <AMReX_AmrCore.H>
+
+//Sets the decorator for GPU
```
In the QED library, all the methods which could conceivably be run on a GPU have a PXRMP_GPU decorator. The user is responsible for setting this parameter to the right value.

Now the QED library can be used either with SI units or with normalized units (c=1, m_e=1, omega=1...).
Units are chosen with a define (PXRMP_WITH_SI_UNITS in this case).
Actually, I am not completely satisfied with this solution. In the future I would like to propose a template-based alternatives, like breit_wheeler_engine<double,units_si> and breit_wheeler_engine<double,units_norm>.

lucafedeli88 replied:
>[...] do we really need all of them to be __host__ __device__, i.e., compiled to run on both the CPU and GPU?
Actually no.  The functions decorated  with PXRMP_GPU will only be called in an `amrex::ParallelFor` cycle . I've just corrected this. 

> Ok, I am find with both solutions, i.e., template or #define (I agree the template one is nicer, but I don't think there is a rush here).

I agree. We could discuss that after the coupling is finished.

lucafedeli88 commented on Source/QED/breit_wheeler_engine_wrapper.h:
```diff
@@ -0,0 +1,27 @@
+
+class warpx_breit_wheeler_engine :
+    public picsar::multi_physics::
+    breit_wheeler_engine<amrex::Real, amrex_rng_wrapper>
```
The idea is that I would like to be able to add (maybe radical) changes to the QED library without affecting too much the WarpX code. For example, if in a few months we switch to a template-based solution to choose between SI units and normalized units, using a wrapper would allow us to change just a couple of files in the WarpX code. 

Instead of creating a derived class, I could put just pure functions  in the QED folder  (sorted in different files). For example, a warpx_breit_wheeler_engine.hpp file could contain:
- definition of PXRMP_GPU & PXRMP_WITH_SI_UNITS
- inclusion of the relevant files of the QED library
- pure functions wrapping the calls to the static functions of the QED libraries
- a pure function which initializes the lookup tables instantiating a BW object

what do you think?

lucafedeli88 commented on Source/QED/breit_wheeler_engine_wrapper.h:
```diff
@@ -0,0 +1,27 @@
+#define WARPX_breit_wheeler_engine_wrapper_h_
+
+//This file provides a wrapper aroud the breit_wheeler engine
+//provided by the standard template library
```
Oh sorry, this is a misprint!
It has been corrected

lucafedeli88 commented on Source/QED/breit_wheeler_engine_wrapper.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "amrex_rng_wrapper.h"
+
+class warpx_breit_wheeler_engine :
```
This has to be decides (see my reply to your comment "Just checking: it seems that class ...")

lucafedeli88 commented on Source/QED/quantum_sync_engine_wrapper.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "amrex_rng_wrapper.h"
+
+class warpx_quantum_sync_engine :
```
This has to be decided (see my reply to your comment "Just checking: it seems that class ...")

lucafedeli88 commented:
Concerning your comments:

> Is this PR actually ready for review? If so, could remove the tag [WIP] in the PR name when it is ready for re-review? If not, sorry for the early review :).

Yes, it was intended to be reviewed, don't worry.

> For all new inputs arguments (e.g., <species>.do_qed_quantum_sync) could you add an entry in the documentation? If the implementation is in progress, this can be explicitly written in the doc.

Done. It's PR lucafedeli88-qed-doc-inputfile.

> Could you add an automated test for both QED processes? An example could be to verify that tau is properly initialized and written to file (we could check that the initial distribution for tau agrees with theory). You can find many examples, for instance files Examples/Modules/ionization/ionization_analysis.py Examples/Modules/ionization/inputs.rt Examples/Modules/ionization/inputs.bf.rt and Regression/WarpX-tests.ini in PR #260. There are many other tests with different options, let me know if you need some help.

I'll do that. For the moment initializing tau is the only thing that the added code actually does, so I'll test that. I'll try to do that this afternoon.

> Could you use CamelCase instead of snake_case for file and class names? This convention has been added to CONTRIBUTING.md in PR #356.

> Could you re-order #include directives to also comply with PR #356? Sorry, these two points should have been added to CONTRIBUTING.md before.

Of course. As soon as everythin else in this PR is fully addressed I will apply this style change. 

>This is a minor point, but I don't think you need to put pp.query("do_qed_breit_wheeler", do_qed_breit_wheeler); etc. in a #ifdef WARPX_QED region. Putting it outside, you can add a test like
```
#ifndef WARPX_QED
AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
    do_qed_breit_wheeler == 0,
    "ERROR: To use do_qed_breit_wheeler=1, you need to compile with WARPX_QED."
);
#endif
```

That's true. If we decide to keep the compilation flag (see dedicated comment) this is definitively the right thing to do.

lucafedeli88 commented:
I should have implemented all the requested changes.

As agreed, QED engines provided by the PICSAR library are wrapped in "functor factory classes" e.g. :

```
class BreitWheelerGetOpticalDepth
{
public:
    BreitWheelerGetOpticalDepth ()
    {};
    AMREX_GPU_DEVICE
    amrex::Real operator() () const;
};

class BreitWheelerEngine
{
public:
    BreitWheelerEngine ();
    BreitWheelerGetOpticalDepth build_optical_depth_functor ();
};
```

These classes are instantiated in the ```MultiParticleContainer``` and a smart pointer to these objects is given to all the species requiring them. Species use functors like this:
```
void PhysicalParticleContainer::InitTauQuantumSync()
{
    BL_PROFILE("PhysicalParticleContainer::InitTauQuantumSync");
    //Get functor
    auto get_opt = shr_ptr_qs_engine->build_optical_depth_functor();
    //Looping over all the particles
    int num_levels = finestLevel() + 1;
    for (int lev=0; lev < num_levels; ++lev)
        for (WarpXParIter pti(*this, lev); pti.isValid(); ++pti){
            auto taus = pti.GetAttribs(particle_comps["tau"]).dataPtr();
            amrex::ParallelFor(
                pti.numParticles(),
                [=] AMREX_GPU_DEVICE (long i) {
                    taus[i] = get_opt();
                }
                );
    }
}
```


I've added docstrings for the ```BreitWheelerEngine``` class and for the ```QuantumSynchrotronGetOpticalDepth``` class. I've also modified the style to comply with the style rules of the project (i.e. CamelCase  for class names and file names) ).

I've added two unit tests, which should check that the initial distribution of the optical depths is exponential (for both photons and leptons). ** However, at present, these tests require PICSAR to be on the QED branch. Therefore, they are commented out in the regression test file **.  The idea is to merge the QED branch with the master branch of PICSAR very soon (i.e. within days).

lucafedeli88 commented:
Unit tests are failing because the QED modules in PICSAR need a more recent version of the Boost libraries (Bessel functions should be included)

MaxThevenet commented on Source/Particles/PhysicalParticleContainer.cpp:
```diff
@@ -2074,3 +2110,50 @@ PhysicalParticleContainer::buildIonizationMask (const amrex::MFIter& mfi, const
+            amrex::ParallelFor(
+                pti.numParticles(),
+                [=] AMREX_GPU_DEVICE (long i) {
+                    taus[i] = get_opt();
```
When running a simulation with  a moving window (typically done when using boosted frame), particles are continuously injected from one side of the box. This is done by calling `AddPlasma` at each iteration, but `InitTauQuantumSync` would be called only once at the initialization of the simulation. With the current implementation, the particles starting inside the box at t=0 would have a proper `tau` initialization, but the particles that are continuously injected wouldn't (even from the same particle container).

To avoid this issue, we should put the `tau` initialization into `AddPlasma`. Would you have time to do it? I'd say you could mimic what is done for ionization (grep `do_field_ionization` in `AddPlasma`, note that member variables have to be copied to temporary variables to avoid the whole `PhysicalParticleContainer` class to be captured by valued and copied to the GPU, as done for `int loc_ionization_initial_level = ionization_initial_level;` for instance). Let me know if I can help.

lucafedeli88 replied:
Sorry, I didn't know that. I've moved tau initialization for both photons and leptons in AddPlasma (which is part of ```PhysicalParticleContainer```).
It seems to be working correctly...

MaxThevenet commented on Source/Particles/WarpXParticleContainer.H:
```diff
@@ -1,12 +1,17 @@
 #ifndef WARPX_WarpXParticleContainer_H_
 #define WARPX_WarpXParticleContainer_H_
 
+#include <memory>
```
Could you put this one down in the list of `include` (basically last)? Or is there a problem doing that?

lucafedeli88 replied:
Sure! Placing this at the beginning of the ``` include ```  list was a mistake.

MaxThevenet commented on Source/Particles/PhotonParticleContainer.cpp:
```diff
@@ -99,3 +122,26 @@ PhotonParticleContainer::Evolve (int lev,
+
+    //Looping over all the particles
+    int num_levels = finestLevel() + 1;
+    for (int lev=0; lev < num_levels; ++lev)
```
Same thing as for PhysicalParticleContainer. Let me know if you want me to look into it.

MaxThevenet commented on Source/QED/BreitWheelerEngineWrapper.cpp:
```diff
@@ -0,0 +1,28 @@
+// Initialization of the optical depth
+AMREX_GPU_DEVICE
+amrex::Real
+BreitWheelerGetOpticalDepth::operator() () const
```
Shouldn't this be `AMREX_FORCE_INLINE` since this is supposed to be called inside a `ParallelFor` loop?
Also, I'm not sure about this, but shouldn't `WarpXBreitWheelerWrapper::internal_get_optical_depth` be also inlined?

lucafedeli88 replied:
You are right.

I've moved the definition of this functor to the header, which is now an ```.hpp``` file.
The definition of the functor now reads as follows:
```
class BreitWheelerGetOpticalDepth
{
public:
    BreitWheelerGetOpticalDepth ()
    {};

    AMREX_GPU_DEVICE 
    AMREX_FORCE_INLINE
    amrex::Real operator() () const
    {
        return WarpXBreitWheelerWrapper::
            internal_get_optical_depth(amrex::Random());
    }
};
``` 

```internal_get_optical_depth``` should be inlined. Indeed, it is declared as follows:
```
         PXRMP_GPU
         PXRMP_FORCE_INLINE
         static _REAL internal_get_optical_depth(_REAL unf_zero_one_minus_epsi);
```

where ```PXRMP_FORCE_INLINE``` is defined as follows:
```
        #if defined(__CUDA_ARCH__)
          #define PXRMP_FORCE_INLINE __forceinline__
        #elif defined(__INTEL_COMPILER)
          #define PXRMP_FORCE_INLINE inline __attribute__((always_inline))
        #elif defined(__clang__)
          #define PXRMP_FORCE_INLINE inline __attribute__((always_inline))
        #elif defined(__GNUC__)
          #define PXRMP_FORCE_INLINE inline __attribute__((always_inline))
        #elif defined(__ibmxl__)
          #define PXRMP_FORCE_INLINE inline __attribute__((always_inline))
        #else
          #define PXRMP_FORCE_INLINE inline
        #endif
```

In the future it would probably be better to provide in PICSAR a way to override the ```PXRMP_FORCE_INLINE``` macro (e.g. ```#define PXRMP_FORCE_INLINE AMREX_FORCE_INLINE```).

MaxThevenet commented on Source/QED/QuantumSyncEngineWrapper.h:
```diff
@@ -0,0 +1,44 @@
+    {};
+
+    AMREX_GPU_DEVICE
+    amrex::Real operator() () const;
```
Same thing, shouldn't this operator be inlined?

lucafedeli88 replied:
You are right. I've addressed this point as above.

MaxThevenet commented on Source/QED/QuantumSyncEngineWrapper.h:
```diff
@@ -0,0 +1,44 @@
+
+//============================================
+
+#endif //WARPX_quantum_sync_engine_wrapper_h_
```
Could you add a newline at the end of the file? GitHub shows a red sign for it...

lucafedeli88 replied:
Ok

MaxThevenet commented:
Just for the record, as you mentioned the tests crash, see error below. We'll fix that before merging.

```
Traceback (most recent call last):
724  File "./check_2d_tau_init.py", line 35, in <module>
725    main()
726  File "./check_2d_tau_init.py", line 32, in main
727    check()
728  File "./check_2d_tau_init.py", line 19, in check
729    res_ele_tau = all_data["electrons", 'particle_tau']
730  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 248, in __getitem__
731    f = self._determine_fields([key])[0]
732  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1159, in _determine_fields
733    finfo = self.ds._get_field_info(ftype, fname)
734  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/static_output.py", line 798, in _get_field_info
735    raise YTFieldNotFound((ftype, fname), self)
736yt.utilities.exceptions.YTFieldNotFound: Could not find field '('electrons', 'particle_tau')' in photon_pusher_plt00050.
737
738   execution time: 10.2s
739   creating problem test report ...
740   photon_pusher FAILED
```

lucafedeli88 commented on Examples/Modules/qed/quantum_synchrotron/inputs.2d_test_tau_init:
```diff
@@ -0,0 +1,92 @@
+warpx.plot_crsepatch = 0
+warpx.use_filter = 1
+warpx.cfl = 1. # if 1., the time step is set to its CFL limit
+warpx.do_pml = 1 # use Perfectly Matched Layer as boundary condition
```
Sure!

lucafedeli88 commented:
> Just for the record, as you mentioned the tests crash, see error below. We'll fix that before merging.
> 
> ```
> Traceback (most recent call last):
> 724  File "./check_2d_tau_init.py", line 35, in <module>
> 725    main()
> 726  File "./check_2d_tau_init.py", line 32, in main
> 727    check()
> 728  File "./check_2d_tau_init.py", line 19, in check
> 729    res_ele_tau = all_data["electrons", 'particle_tau']
> 730  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 248, in __getitem__
> 731    f = self._determine_fields([key])[0]
> 732  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/data_containers.py", line 1159, in _determine_fields
> 733    finfo = self.ds._get_field_info(ftype, fname)
> 734  File "/home/travis/miniconda3/lib/python3.7/site-packages/yt/data_objects/static_output.py", line 798, in _get_field_info
> 735    raise YTFieldNotFound((ftype, fname), self)
> 736yt.utilities.exceptions.YTFieldNotFound: Could not find field '('electrons', 'particle_tau')' in photon_pusher_plt00050.
> 737
> 738   execution time: 10.2s
> 739   creating problem test report ...
> 740   photon_pusher FAILED
> ```

The new ```.travis.yml``` file installs the required dependencies.

MaxThevenet commented on Source/QED/BreitWheelerEngineWrapper.hpp:
```diff
@@ -0,0 +1,49 @@
+
+// Factory class =============================
+
+/* \brief Wrapper for the Breit Wheeler engine of the PICSAR library */
```
Ah one last comment. I am not very familiar with Doxygen, so correct me if I'm wrong, but is
```/* \brief Wrapper for the Breit Wheeler engine of the PICSAR library */```
a proper Doxygen formatting? Shoudn't it be e.g.,
```/** \brief Wrapper for the Breit Wheeler engine of the PICSAR library */```
?

MaxThevenet replied:
Have you had a change to check that?

lucafedeli88 replied:
It _should_ work now, but I am not familiar with Doxygen too...

MaxThevenet commented:
Looks good to me, thanks! I'll merge when tests pass.
